src/Plutip/UtxoDistribution.purs

```diff
@@ -0,0 +1,184 @@
+module Plutip.UtxoDistribution
+  ( class UtxoDistribution
+  , decodeWallets
+  , decodeWallets'
+  , keyWallets
+  , encodeDistribution
+  , transferFundsFromEnterpriseToBase
+  , withStakeKey
+  ) where
+
+import Prelude
+
+import Contract.Address
+  ( PaymentPubKeyHash
+  , StakePubKeyHash
+  , getNetworkId
+  , getWalletAddress
+  , ownPaymentPubKeyHash
+  , ownStakePubKeyHash
+  , payPubKeyHashEnterpriseAddress
+  )
+import Contract.Monad (Contract, liftContractM, liftedE, liftedM)
+import Contract.Prelude (foldM, foldMap, null)
+import Contract.ScriptLookups as Lookups
+import Contract.Transaction
+  ( TransactionOutput(TransactionOutput)
+  , awaitTxConfirmed
+  , balanceAndSignTxE
+  , signTransaction
+  , submit
+  )
+import Contract.TxConstraints as Constraints
+import Contract.Utxos (utxosAt)
+import Contract.Wallet (withKeyWallet)
+import Data.Array as Array
+import Data.FoldableWithIndex (foldMapWithIndex)
+import Data.List (List, (:))
+import Data.Maybe (Maybe(Nothing, Just))
+import Data.Newtype (unwrap, wrap)
+import Data.Tuple (fst)
+import Data.Tuple.Nested (type (/\), (/\))
+import Plutip.Types
+  ( InitialUTxO
+  , InitialUTxOWithStakeKey(InitialUTxOWithStakeKey)
+  , PrivateKeyResponse(PrivateKeyResponse)
+  , UtxoAmount
+  )
+import Plutus.Types.Transaction (Utxo)
+import Type.Prelude (Proxy(Proxy))
+import Wallet.Key
+  ( KeyWallet
+  , PrivatePaymentKey(PrivatePaymentKey)
+  , PrivateStakeKey
+  , privateKeysToKeyWallet
+  )
+
+-- | A type class that implements a type-safe interface for specifying UTXO
+-- | distribution for wallets.
+-- | Number of wallets in distribution specification matches the number of
+-- | wallets provided to the user.
+class UtxoDistribution distr wallets | distr -> wallets where
+  encodeDistribution :: distr -> Array (Array UtxoAmount)
+  decodeWallets :: distr -> Array PrivateKeyResponse -> Maybe wallets
```

Comment:
Should probably add a check in the `decodeWallets` definitions that the final `PrivateKeyResponse` array is empty